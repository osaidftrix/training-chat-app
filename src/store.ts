import { combineReducers, configureStore } from '@reduxjs/toolkit';
import connectionSlice from './store/slices/connectionSlice';
import storage from 'redux-persist/lib/storage';
//import storage from 'redux-persist-indexeddb-storage';
import { persistReducer, persistStore } from 'redux-persist';
import thunk from 'redux-thunk';
const persistConfig = {
  key: 'root',
  storage,
  blacklist: ['connection'],
  // whitelist: ['connection']
};

const rootReducer = combineReducers({
  connection: connectionSlice,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: [thunk],
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
