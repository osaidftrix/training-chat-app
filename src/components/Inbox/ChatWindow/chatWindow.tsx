import { useForm, SubmitHandler } from 'react-hook-form';
import { useAppSelector } from '../../../store/hooks';
import { IMessage, getConnection, setMessages } from '../../../store/slices/connectionSlice';
import Message from '../Message/message';
import { useDispatch } from 'react-redux';
import React, { useEffect, useMemo, useRef } from 'react';

export interface IChatWindow {
  toggledUser: string;
  setToggledUser: React.Dispatch<React.SetStateAction<string>>;
}

const ChatWindow: React.FC<IChatWindow> = (props) => {
  const dispatch = useDispatch();
  const ref: React.RefObject<HTMLDivElement> = useRef(null);
  const {
    register,
    handleSubmit,
    reset,
    // watch
    // formState: { errors },
  } = useForm<{ input: string }>();
  const onSubmit: SubmitHandler<{ input: string }> = async (data) => {
    if (!data.input) return;
    try {
      const message: IMessage = {
        username: connection.userName as string,
        message: data.input,
      };
      let response;
      if (props.toggledUser === 'public') {
        response = await connection.connectionInstance?.invoke('BroadcastChat', message);
      } else {
        response = await connection.connectionInstance?.invoke('SendChatToUser', props.toggledUser, message);
        if (response && props.toggledUser !== connection.userName) {
          dispatch(setMessages({ target: props.toggledUser, message }));
        }
        console.log('invoked message status: ', response);
        console.log('message params', message, props.toggledUser);
      }
    } catch (err) {
      console.log('err', err);
    }

    reset();
  };
  const connection = useAppSelector(getConnection);
  const messages = useMemo(() => connection.messages[props.toggledUser] ?? [], [connection.messages, props.toggledUser]);

  useEffect(() => {
    if (ref.current) {
      ref.current.scrollIntoView({ behavior: 'smooth' });
    }
  }, [messages]);

  return (
    <>
      <div className="w-full h-[92vh] px-5 flex flex-col justify-between">
        <div className="flex flex-col mt-5 overflow-y-auto">
          {messages.map((message, index) => (
            <Message key={`${message.username}-${index}`} {...message} {...props} />
          ))}
          <div ref={ref} />
        </div>
        <div className="py-5">
          <form onSubmit={handleSubmit(onSubmit)}>
            <label htmlFor="chat" className="sr-only">
              Your message
            </label>
            <div className="flex items-center px-3 py-2 rounded-lg bg-gray-50">
              <textarea
                {...register('input')}
                rows={1}
                className="block mx-4 p-2.5 w-full text-sm text-gray-900 bg-white rounded-lg border border-gray-300 focus:ring-blue-500 focus:border-blue-500"
                placeholder="Your message..."
                onKeyDown={(e) => {
                  if (e.code === 'Enter') {
                    e.preventDefault();
                    handleSubmit(onSubmit)();
                  }
                }}
              />
              <button
                type="submit"
                className="inline-flex justify-center p-2 text-blue-600 rounded-full cursor-pointer hover:bg-blue-100"
              >
                <svg
                  className="w-5 h-5 rotate-90"
                  aria-hidden="true"
                  xmlns="http://www.w3.org/2000/svg"
                  fill="currentColor"
                  viewBox="0 0 18 20"
                >
                  <path d="m17.914 18.594-8-18a1 1 0 0 0-1.828 0l-8 18a1 1 0 0 0 1.157 1.376L8 18.281V9a1 1 0 0 1 2 0v9.281l6.758 1.689a1 1 0 0 0 1.156-1.376Z" />
                </svg>
                <span className="sr-only">Send message</span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </>
  );
};

export default ChatWindow;
