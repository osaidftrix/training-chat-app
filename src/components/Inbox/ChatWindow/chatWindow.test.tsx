import { describe, test, expect } from 'vitest';
import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';
import { PersistGate } from 'redux-persist/integration/react';
import { persistor, store } from '../../../store';
import ChatWindow from './chatWindow';

describe('<ChatWindow />', () => {
  test('ChatWindow mounts properly', () => {
    const renderer = create(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <ChatWindow toggledUser={'testuser'} setToggledUser={() => {}} />
        </PersistGate>
      </Provider>,
    );

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
