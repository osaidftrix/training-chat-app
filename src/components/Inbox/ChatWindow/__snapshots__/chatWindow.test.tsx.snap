// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ChatWindow /> > ChatWindow mounts properly 1`] = `
<div
  className="w-full h-[92vh] px-5 flex flex-col justify-between"
>
  <div
    className="flex flex-col mt-5 overflow-y-auto"
  >
    <div />
  </div>
  <div
    className="py-5"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        className="sr-only"
        htmlFor="chat"
      >
        Your message
      </label>
      <div
        className="flex items-center px-3 py-2 rounded-lg bg-gray-50"
      >
        <textarea
          className="block mx-4 p-2.5 w-full text-sm text-gray-900 bg-white rounded-lg border border-gray-300 focus:ring-blue-500 focus:border-blue-500"
          name="input"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Your message..."
          rows={1}
        />
        <button
          className="inline-flex justify-center p-2 text-blue-600 rounded-full cursor-pointer hover:bg-blue-100"
          type="submit"
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5 rotate-90"
            fill="currentColor"
            viewBox="0 0 18 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m17.914 18.594-8-18a1 1 0 0 0-1.828 0l-8 18a1 1 0 0 0 1.157 1.376L8 18.281V9a1 1 0 0 1 2 0v9.281l6.758 1.689a1 1 0 0 0 1.156-1.376Z"
            />
          </svg>
          <span
            className="sr-only"
          >
            Send message
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<UserList /> > UserList mounts properly 1`] = `
<div
  className="w-full h-[92vh] px-5 flex flex-col justify-between"
>
  <div
    className="flex flex-col mt-5 overflow-y-auto"
  >
    <div />
  </div>
  <div
    className="py-5"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        className="sr-only"
        htmlFor="chat"
      >
        Your message
      </label>
      <div
        className="flex items-center px-3 py-2 rounded-lg bg-gray-50"
      >
        <textarea
          className="block mx-4 p-2.5 w-full text-sm text-gray-900 bg-white rounded-lg border border-gray-300 focus:ring-blue-500 focus:border-blue-500"
          name="input"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Your message..."
          rows={1}
        />
        <button
          className="inline-flex justify-center p-2 text-blue-600 rounded-full cursor-pointer hover:bg-blue-100"
          type="submit"
        >
          <svg
            aria-hidden="true"
            className="w-5 h-5 rotate-90"
            fill="currentColor"
            viewBox="0 0 18 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m17.914 18.594-8-18a1 1 0 0 0-1.828 0l-8 18a1 1 0 0 0 1.157 1.376L8 18.281V9a1 1 0 0 1 2 0v9.281l6.758 1.689a1 1 0 0 0 1.156-1.376Z"
            />
          </svg>
          <span
            className="sr-only"
          >
            Send message
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
